presto-main/src/main/java/io/prestosql/operator/scalar/BitwiseFunctions.java

```diff
@@ -75,4 +75,115 @@ public static long bitwiseXor(@SqlType(StandardTypes.BIGINT) long left, @SqlType
     {
         return left ^ right;
     }
+
+    @Description("bitwise left shift")
+    @ScalarFunction("bitwise_left_shift")
+    @SqlType(StandardTypes.TINYINT)
+    public static long bitwiseLeftShiftTinyint(@SqlType(StandardTypes.TINYINT) long num, @SqlType(StandardTypes.INTEGER) long shift)
+    {
+        return bitwiseLeftShift(num, shift);
```

Comment:
This needs to mask the bottom 8 bits

Reply:
It's also more involved for negative numbers. For instance, TINYINT `-2` is represented as `111....11111110`. However, when shifting, we can't blindly mask the bottom bits.

We also need to think about what the behavior should be if the shift is larger than the number of bits in the type. For instance, what should `bitwise_left_shift(BIGINT '1', 70)`, do? In Java, it's a no-op, but I find that confusing and not consistent.

Reply:
In the original suggestion, we specify the digit controlling how we treat the input. Similarly, we may be able to mask the input with 8, 16, and 32 bits and do the shift operation. That seems to be straightforward. 
If the shift is larger than the bits in the type, how about treating it as 0? Thus `bitwise_left_shift(BIGINT '1', 70)` will be the long type with all bits 0. For the arithmetic shift, we keep the signed bit but other bits are 0 because they are all shifted.

Reply:
I found simply masking with eight 1s is not working. The returned type is expected to be signed `TINYINT`. But the returned type is unsigned value, which caused the overflow of the value.  For instance, `-7 << 2` should return -28 (11100100) but it's 228 if the head bit is not interpreted as the sign. Returning 11100100 as `TINYINT` fails. 

```
io.prestosql.spi.PrestoException: Cannot cast '228' to TINYINT
```

How can we treat the returned value as a signed value?

Reply:
I think masking the value seems to cause a confusing result. Even if the result is within the valid range (e.g. -7 << 2 = -28 signed byte), masking gets rid of signs and throws the out of range error. It causes a vast amount of inconsistent results with the Java implementation.

We may only need to check the range of the value whether it's within the result type or not as follows. The outcome looks like more straightforward and consistent with Java's result.

```java
    public static long bitwiseLeftShiftTinyint(@SqlType(StandardTypes.TINYINT) long num, @SqlType(StandardTypes.INTEGER) long shift)
    {
        return SignedBytes.checkedCast(num << shift);
    }
```

Reply:
We need to mask while preserving the sign. Basically, for INTEGER, the value should be have as if it were a 32-bit signed value; for SMALLINT, as if it were a 16-bit signed value; for TINYINT, as if it were an 8-bit signed value. Given that they are stored in a `long`, we need to perform the shift such as if the sign bit of the 32, 16 or 8 bit value would be set, the sign bit of the underlying `long` is set and produce the same numeric value as if it was dealing with the the shorter data types.

For instance, 

```
TINYINT '1' << 7 ==> -128
TINYINT '-128' << 1 ==> 0 (the sign bit drops off the left side)
TINYINT '-65' << 1 ==> 130 (this is 1011_1111 << 1, which should produce 0111_1110)
```

Reply:
Thanks for giving me examples. I tried to understand what the functions should behave. If I understand correctly, 

1. Shift values as it is
2. If the signed bit is active, pad the upper bits with 1 to get the correct signed number in `long`.
3. If the signed bit is inactive, mask the lower bits with the size of the specified type.

But the following example shows different results.
> TINYINT '-65' << 1 ==> 130 (this is 1011_1111 << 1, which should produce 0111_1110)

This operation produces 126 (`0111_1110`) if it's interpreted as a signed 8-bit number. 130 is out of the range of signed 8-bit numbers.

Reply:
Yes, you're right. I mistyped the number. It should be 126.